// command-line-interface/build/lib/runCli.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.runCli = void 0;
const addHelpCommandToCli_1 = require("./addHelpCommandToCli");
const getGetUsage_1 = require("./usage/getGetUsage");
const getRecommendCommand_1 = require("./recommend/getRecommendCommand");
const runCliRecursive_1 = require("./runCliRecursive");
const runCli = async function ({ rootCommand, argv, handlers = {} }) {
    const handlersWithDefaults = {
        commandFailed({ ex }) {
            // eslint-disable-next-line no-console
            console.error(ex);
        },
        commandUnknown({ unknownCommandName, recommendedCommandName }) {
            // eslint-disable-next-line no-console
            console.error(`Unknown command '${unknownCommandName}'. Did you mean '${recommendedCommandName}'?`);
        },
        optionInvalid({ reason }) {
            // eslint-disable-next-line no-console
            console.error(reason);
        },
        optionMissing({ optionDefinition }) {
            // eslint-disable-next-line no-console
            console.error(`Option '${optionDefinition.name}' is missing.`);
        },
        optionUnknown({ optionName }) {
            // eslint-disable-next-line no-console
            console.error(`Unknown option '${optionName}'.`);
        },
        ...handlers
    };
    const extendedRootCommand = addHelpCommandToCli_1.addHelpCommandToCli({ rootCommand });
    const recommendCommand = getRecommendCommand_1.getRecommendCommand({ rootCommand: extendedRootCommand });
    const getUsage = getGetUsage_1.getGetUsage({ rootCommand: extendedRootCommand });
    await runCliRecursive_1.runCliRecursive({
        command: extendedRootCommand,
        argv,
        handlers: handlersWithDefaults,
        getUsage,
        recommendCommand,
        level: 0,
        ancestorOptions: {},
        ancestorNames: []
    });
};
exports.runCli = runCli;
