// defekt/build/lib/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.value = exports.isResult = exports.isError = exports.isCustomError = exports.error = exports.defekt = exports.CustomError = void 0;
const CustomError_1 = require("./CustomError");
Object.defineProperty(exports, "CustomError", { enumerable: true, get: function () { return CustomError_1.CustomError; } });
const defekt_1 = require("./defekt");
Object.defineProperty(exports, "defekt", { enumerable: true, get: function () { return defekt_1.defekt; } });
const isCustomError_1 = require("./isCustomError");
Object.defineProperty(exports, "isCustomError", { enumerable: true, get: function () { return isCustomError_1.isCustomError; } });
const isError_1 = require("./isError");
Object.defineProperty(exports, "isError", { enumerable: true, get: function () { return isError_1.isError; } });
const isResult_1 = require("./isResult");
Object.defineProperty(exports, "isResult", { enumerable: true, get: function () { return isResult_1.isResult; } });
const Result_1 = require("./Result");
Object.defineProperty(exports, "error", { enumerable: true, get: function () { return Result_1.error; } });
Object.defineProperty(exports, "value", { enumerable: true, get: function () { return Result_1.value; } });
