// typedescriptor/build/lib/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.typeOf = exports.isUndefined = exports.isSymbol = exports.isString = exports.isSet = exports.isScalar = exports.isReference = exports.isObject = exports.isNumber = exports.isNull = exports.isMap = exports.isFunction = exports.isError = exports.isBoolean = exports.isArray = void 0;
const isArray_1 = require("./isArray");
Object.defineProperty(exports, "isArray", { enumerable: true, get: function () { return isArray_1.isArray; } });
const isBoolean_1 = require("./isBoolean");
Object.defineProperty(exports, "isBoolean", { enumerable: true, get: function () { return isBoolean_1.isBoolean; } });
const defekt_1 = require("defekt");
Object.defineProperty(exports, "isError", { enumerable: true, get: function () { return defekt_1.isError; } });
const isFunction_1 = require("./isFunction");
Object.defineProperty(exports, "isFunction", { enumerable: true, get: function () { return isFunction_1.isFunction; } });
const isMap_1 = require("./isMap");
Object.defineProperty(exports, "isMap", { enumerable: true, get: function () { return isMap_1.isMap; } });
const isNull_1 = require("./isNull");
Object.defineProperty(exports, "isNull", { enumerable: true, get: function () { return isNull_1.isNull; } });
const isNumber_1 = require("./isNumber");
Object.defineProperty(exports, "isNumber", { enumerable: true, get: function () { return isNumber_1.isNumber; } });
const isObject_1 = require("./isObject");
Object.defineProperty(exports, "isObject", { enumerable: true, get: function () { return isObject_1.isObject; } });
const isReference_1 = require("./isReference");
Object.defineProperty(exports, "isReference", { enumerable: true, get: function () { return isReference_1.isReference; } });
const isScalar_1 = require("./isScalar");
Object.defineProperty(exports, "isScalar", { enumerable: true, get: function () { return isScalar_1.isScalar; } });
const isSet_1 = require("./isSet");
Object.defineProperty(exports, "isSet", { enumerable: true, get: function () { return isSet_1.isSet; } });
const isString_1 = require("./isString");
Object.defineProperty(exports, "isString", { enumerable: true, get: function () { return isString_1.isString; } });
const isSymbol_1 = require("./isSymbol");
Object.defineProperty(exports, "isSymbol", { enumerable: true, get: function () { return isSymbol_1.isSymbol; } });
const isUndefined_1 = require("./isUndefined");
Object.defineProperty(exports, "isUndefined", { enumerable: true, get: function () { return isUndefined_1.isUndefined; } });
const typeOf_1 = require("./typeOf");
Object.defineProperty(exports, "typeOf", { enumerable: true, get: function () { return typeOf_1.typeOf; } });
