// buntstift/build/lib/Configuration.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Configuration = void 0;
class Configuration {
    constructor({ colorLevel, isInteractiveSession, isQuietModeEnabled, isUtf8Enabled, isVerboseModeEnabled }) {
        this.colorLevel = colorLevel;
        this.isInteractiveSession = isInteractiveSession;
        this.isQuietModeEnabled = isQuietModeEnabled;
        this.isUtf8Enabled = isUtf8Enabled;
        this.isVerboseModeEnabled = isVerboseModeEnabled;
    }
    clone() {
        const clonedConfiguration = new Configuration({
            colorLevel: this.colorLevel,
            isInteractiveSession: this.isInteractiveSession,
            isQuietModeEnabled: this.isQuietModeEnabled,
            isUtf8Enabled: this.isUtf8Enabled,
            isVerboseModeEnabled: this.isVerboseModeEnabled
        });
        return clonedConfiguration;
    }
    withColorLevel(colorLevel) {
        const newConfiguration = this.clone();
        newConfiguration.colorLevel = colorLevel;
        return newConfiguration;
    }
    withInteractiveSession(isInteractiveSession) {
        const newConfiguration = this.clone();
        newConfiguration.isInteractiveSession = isInteractiveSession;
        return newConfiguration;
    }
    withQuietMode(isQuietModeEnabled) {
        const newConfiguration = this.clone();
        newConfiguration.isQuietModeEnabled = isQuietModeEnabled;
        return newConfiguration;
    }
    withUtf8(isUtf8Enabled) {
        const newConfiguration = this.clone();
        newConfiguration.isUtf8Enabled = isUtf8Enabled;
        return newConfiguration;
    }
    withVerboseMode(isVerboseModeEnabled) {
        const newConfiguration = this.clone();
        newConfiguration.isVerboseModeEnabled = isVerboseModeEnabled;
        return newConfiguration;
    }
}
exports.Configuration = Configuration;
